src/Compiler/Facilities/prim-lexing.fs

```diff
@@ -108,6 +110,24 @@ type StringText(str: string) =
         member _.CopyTo(sourceIndex, destination, destinationIndex, count) =
             str.CopyTo(sourceIndex, destination, destinationIndex, count)
 
+        member this.GetSubTextFromRange(range) =
+            let sourceText = this :> ISourceText
+            let startLine = range.StartLine - 1
+            let line = sourceText.GetLineString startLine
+
+            if range.StartLine = range.EndLine then
+                let length = range.EndColumn - range.StartColumn
+                line.Substring(range.StartColumn, length)
+            else
+                let firstLineContent = line.Substring(range.StartColumn)
+                let sb = System.Text.StringBuilder().AppendLine(firstLineContent)
+
+                (sb, [ range.StartLine .. range.EndLine - 2 ])
+                ||> List.fold (fun sb lineNumber -> sb.AppendLine(sourceText.GetLineString lineNumber))
+                |> fun sb ->
+                    let lastLine = sourceText.GetLineString(range.EndLine - 1)
+                    sb.Append(lastLine.Substring(0, range.EndColumn)).ToString()
+
```

Comment:
This looks like it allocates a lot. Probably worth it to start introducing span-based APIs.

Reply:
I am all ears on how to do that stuff 😉!

Reply:
I'll tinker with it tomorrow

Reply:
@vzarytovskii, apropos what I said earlier
https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=net-7.0#system-text-stringbuilder-append(system-readonlymemory((system-char)))

Core 3+. Sad!